packages/flame/lib/src/components/mixins/has_paint.dart

```diff
@@ -97,4 +100,91 @@ mixin HasPaint<T extends Object> on Component {
   void tint(Color color, {T? paintId}) {
     getPaint(paintId).colorFilter = ColorFilter.mode(color, BlendMode.srcATop);
   }
+
+  @override
+  double get opacity => paint.color.opacity;
+
+  @override
+  set opacity(double value) {
+    paint.color = paint.color.withOpacity(value);
+    for (final paint in _paints.values) {
+      paint.color = paint.color.withOpacity(value);
+    }
+  }
+
+  /// Creates an [OpacityProvider] for given [paintId] and can be used as
+  /// `target` for [OpacityEffect].
+  OpacityProvider opacityProviderOf(T paintId) {
+    return _ProxyOpacityProvider(paintId, this);
+  }
+
+  /// Creates an [OpacityProvider] for given list of [paintIds] and can be
+  /// used as `target` for [OpacityEffect].
+  ///
+  /// When opacities of all the given [paintIds] are not same, this provider
+  /// directly effects opacity of the most opaque paint. Additionally, it
+  /// modifies other paints such that their respective opacity ratio with most
+  /// opaque paint is maintained.
+  ///
+  /// If [paintIds] is null or empty, all the paints are used for creating the
+  /// [OpacityProvider].
+  ///
+  /// Note: Each call results in a new [OpacityProvider] and hence the cached
+  /// opacity ratios are calculated using opacities when this method was called.
+  OpacityProvider opacityProviderOfList({List<T?>? paintIds}) {
+    return _MultiPaintOpacityProvider(
+      paintIds ?? (List<T?>.from(_paints.keys)..add(null)),
+      this,
+    );
+  }
+}
+
+class _ProxyOpacityProvider<T extends Object> implements OpacityProvider {
+  _ProxyOpacityProvider(this.paintId, this.target);
+
+  final T paintId;
+  final HasPaint<T> target;
+
+  @override
+  double get opacity => target.getOpacity(paintId: paintId);
+
+  @override
+  set opacity(double value) => target.setOpacity(value, paintId: paintId);
+}
+
+class _MultiPaintOpacityProvider<T extends Object> implements OpacityProvider {
```

Comment:
Where/how is this used?

Reply:
`opacityProviderOfList()` uses this. Using that method, users can specify a list of paintIds on which the provider should operate.